Samples/ProjectTracker/ProjectTracker.Ui.Blazor.Server/Pages/ProjectEdit.razor

```diff
@@ -0,0 +1,229 @@
+          </tr>
+        </tbody>
+      </table>
+    }
+    else if (viewMode == SubViewModes.Select)
+    {
+      @if (_resourceList == null)
+      {
+        <p>Loading resource list...</p>
+      }
+      else
+      {
+        <a href="javascript: void(0);" @onclick="ResetResourcesView">Cancel assignment</a>
+        <table class="table">
+          <thead>
+            <tr>
+              <th>Resource</th>
+              <th></th>
+            </tr>
+          </thead>
+          <tbody>
+            @foreach (var item in _resourceList)
+            {
+              <tr>
+                <td>@item.Name</td>
+                <td><a href="javascript: void(0);" @onclick="() => AssignRole(item.Id)">Select</a></td>
+              </tr>
+            }
+          </tbody>
+        </table>
+      }
+    }
+    else if (viewMode == SubViewModes.Default)
+    {
+      <a href="javascript: void(0);" @onclick="SelectResource">Add resource</a>
+      <table class="table">
+        <thead>
+          <tr>
+            <th>First name</th>
+            <th>Last name</th>
+            <th>Role</th>
+            <th></th>
+          </tr>
+        </thead>
+        <tbody>
+          @foreach (var item in vm.Model.Resources)
+          {
+            <tr>
+              <td>@item.FirstName</td>
+              <td>@item.LastName</td>
+              <td>@item.RoleName</td>
+              <td><a href="javascript: void(0);" @onclick="() => EditResource(item.ResourceId)">Edit</a> | <a href="javascript: void(0);" @onclick="() => RemoveResource(item.ResourceId)">Remove</a></td>
+            </tr>
+          }
+        </tbody>
+      </table>
+    }
+    else
+    {
+      <p>Error in page - please contact support</p>
+    }
+  </div>
+  <button @onclick="vm.SaveAsync" disabled="@(!vm.Model.IsSavable)">Save</button>
+}
+
+@code {
+  [Parameter]
+  public string id { get; set; }
+
+  private SubViewModes viewMode = SubViewModes.Default;
+  private ProjectTracker.Library.ProjectResourceEdit selectedResource;
+  private List<ProjectTracker.Library.ResourceInfo> _resourceList;
+
+  protected override void OnInitialized()
+  {
+    vm.Saved += () => NavigationManager.NavigateTo("projects");
+    vm.ModelChanging += (o, n) =>
+    {
+      if (o != null)
+        o.PropertyChanged -= async (s, e) => await InvokeAsync(() => StateHasChanged());
+      if (n != null)
+        n.PropertyChanged += async (s, e) => await InvokeAsync(() => StateHasChanged());
+    };
+  }
+
+  protected override async Task OnParametersSetAsync()
+  {
+    if (string.IsNullOrWhiteSpace(id))
+      await vm.RefreshAsync();
+    else
+      await vm.RefreshAsync(int.Parse(id));
+  }
+
+  private async void SelectResource()
+  {
+    viewMode = SubViewModes.Select;
+    _resourceList = (await ProjectTracker.Library.ResourceList.GetResourceListAsync())
+                      .Where(r => !vm.Model.Resources.Contains(r.Id)).ToList();
+    StateHasChanged();
```

Comment:
I wouldn't have expected that StateHasChanged was needed in this sort of scenario. This method is being triggered by an event handler in the same component. I thought it was only needed when a child component raises an event that requires a parent component to re-render itself?

It shouldn't do any harm, but it might be a line of code that isn't needed, that's all.